public/service-worker.js

```diff
@@ -0,0 +1,253 @@
+      })(),
+    );
+  }
+});
+
+// Occurs when the service worker is installed
+// https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/install_event
+self.addEventListener("install", (event) => {
+  event.waitUntil(
+    (async () => {
+      const cache = await caches.open(OFFLINE_CACHE);
+      await cache.addAll(CACHE_STATIC_FILES);
+      // Force the newest service worker to take over
+      // https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/skipWaiting
+      self.skipWaiting();
+    })(),
+  );
+});
+
+// Occurs when the service worker takes control of the page
+// https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/activate_event
+self.addEventListener("activate", (event) => {
+  event.waitUntil(
+    (async () => {
+      if ("navigationPreload" in self.registration) {
+        await self.registration.navigationPreload.enable();
+      }
+      const cacheKeys = await caches.keys();
+      await Promise.all(
+        cacheKeys.map((key) => {
+          if (
+            key !== OFFLINE_CACHE &&
+            key !== DATA_CACHE &&
+            key !== NOTIFICATION_CACHE
+          ) {
+            return caches.delete(key);
+          }
+        }),
+      );
+    })(),
+  );
+  self.clients.claim();
+});
+
+// Occurs whenever the main app thread makes a network request
+// https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/fetch_event
+self.addEventListener("fetch", (event) => {
+  const requestUrl = new URL(event.request.url);
+  if (event.request.mode === "navigate") {
+    handleNavigationRequest(event);
+  } else if (CACHE_STATIC_FILES.includes(requestUrl.pathname)) {
+    event.respondWith(
+      caches.match(event.request).then((response) => {
+        return (
+          response ||
+          fetch(event.request).then((networkResponse) => {
+            return caches.open(OFFLINE_CACHE).then((cache) => {
+              cache.put(event.request, networkResponse.clone());
+              return networkResponse;
+            });
+          })
+        );
+      }),
+    );
+    // Tries to serve the requested data from the DATA_CACHE
+    // If unable to, then it makes a network fetch and stores the data in the cache
+  } else if (requestUrl.pathname.startsWith("/bookmarks_in_pipeline")) {
+    event.respondWith(
+      caches.open(DATA_CACHE).then((cache) =>
+        cache.match(event.request, { ignoreSearch: true }).then((response) => {
+          if (response) {
+            return response;
+          } else {
+            return fetch(event.request).then((networkResponse) => {
+              cache.put(event.request, networkResponse.clone());
+              return networkResponse;
+            });
+          }
+        }),
+      ),
+    );
+  }
+});
+
+// Returns a preloaded response if it is enabled.
+// Otherwise it makes a network fetch and if that fails, it returns the offline fallback page
+function handleNavigationRequest(event) {
```

Comment:
How is the cache going to interact with our pages. For example, if I translated some words the URL stays the same when is it that it's refreshed?

Reply:
So the OFFLINE_CACHE in charge of the offline page is opened and collects the static files when the service worker is installed.

The DATA_CACHE, which saves the user's words and their translation, collects this whenever the ("/bookmarks_in_pipeline") endpoint is fetched.